src/thunderbots/software/util/visualizer_messenger/visualizer_messenger.cpp

```diff
@@ -11,20 +20,55 @@ namespace Util
         return vm;
     }
 
-    void VisualizerMessenger::initializePublisher(ros::NodeHandle node_handle)
+    void VisualizerMessenger::onConnection()
     {
-        if (this->publisher)
+        // The io_context is required for all I/O
+        boost::asio::io_service ioc{1};
+
+        // The acceptor receives incoming connections
+        // TODO: make IP and port a configurable constant somewhere?
+        auto const address = boost::asio::ip::address::from_string("127.0.0.1");
+        auto const port    = static_cast<unsigned short>(std::atoi("9091"));
+        tcp::acceptor acceptor{ioc, {address, port}};
+        for (;;)
         {
-            LOG(WARNING) << "Re-initializing visualizer messenger singleton publisher";
-        }
+            // This will receive the new connection
+            tcp::socket socket{ioc};
+
+            // Block until we get a connection
+            acceptor.accept(socket);
+
+            // Move the connection to a websocket stream
+            websocket::stream<tcp::socket> websocket(std::move(socket));
+
+            // Accept the websocket handshake
+            websocket.accept();
+
+            // Set the websocket to talk in binary
+            websocket.binary(true);
+
+            LOG(INFO) << "Connection received." << std::endl;
+
+            // Lock the current list of sockets
+            ws_mutex.lock();
+
+            // Save this new websocket connection
+            ws_connections.emplace_back(std::move(websocket));
+
+            // Unlock the current list of sockets
+            ws_mutex.unlock();
 
-        this->publisher = node_handle.advertise<LayerMsg>(
-            Util::Constants::VISUALIZER_DRAW_LAYER_TOPIC, 8);
+            LOG(INFO) << "Connection added." << std::endl;
+
+            // TODO: this is disgusting. We should not tie up all the cpu
+            // Yield so that other things can run
+            std::this_thread::yield();
```

Comment:
I'm pretty sure we don't need this, because we block and yield when we call `acceptor.accept(socket);`

Reply:
Removed